illumos-utils/src/opte/port_manager.rs

```diff
@@ -125,13 +131,206 @@ impl PortManagerInner {
+    // for the external and VPC-private addresses match.
+    //
+    // See https://github.com/oxidecomputer/omicron/issues/9318.
     pub source_nat: Option<SourceNatConfig>,
     pub ephemeral_ip: Option<IpAddr>,
     pub floating_ips: &'a [IpAddr],
     pub firewall_rules: &'a [ResolvedVpcFirewallRule],
     pub dhcp_config: DhcpCfg,
 }
 
+impl<'a> TryFrom<&PortCreateParams<'a>> for IpCfg {
+    type Error = Error;
+
+    fn try_from(params: &PortCreateParams) -> Result<IpCfg, Error> {
+        match &params.nic.ip_config {
+            PrivateIpConfig::V4(v4) => {
+                build_opte_ipv4_config(
+                    v4,
+                    params.source_nat.as_ref(),
+                    params.ephemeral_ip.as_ref(),
+                    &params.floating_ips,
+                    /* is_ipv4_only = */ true,
+                )
+                .map(IpCfg::Ipv4)
+            }
+            PrivateIpConfig::V6(v6) => {
+                build_opte_ipv6_config(
+                    v6,
+                    params.source_nat.as_ref(),
+                    params.ephemeral_ip.as_ref(),
+                    &params.floating_ips,
+                    /* is_ipv6_only = */ true,
+                )
+                .map(IpCfg::Ipv6)
+            }
+            PrivateIpConfig::DualStack { v4, v6 } => {
+                let ipv4 = build_opte_ipv4_config(
+                    v4,
+                    params.source_nat.as_ref(),
+                    params.ephemeral_ip.as_ref(),
+                    &params.floating_ips,
+                    /* is_ipv4_only = */ false,
+                )?;
+                let ipv6 = build_opte_ipv6_config(
+                    v6,
+                    params.source_nat.as_ref(),
+                    params.ephemeral_ip.as_ref(),
+                    &params.floating_ips,
+                    /* is_ipv6_only = */ false,
+                )?;
+                Ok(IpCfg::DualStack { ipv4, ipv6 })
+            }
+        }
+    }
+}
+
+fn build_opte_ipv4_config(
+    v4: &PrivateIpv4Config,
+    source_nat: Option<&SourceNatConfig>,
+    ephemeral_ip: Option<&IpAddr>,
+    floating_ips: &[IpAddr],
+    is_ipv4_only: bool,
+) -> Result<Ipv4Cfg, Error> {
+    let gateway_ip = v4.subnet().first_host().into();
+    let vpc_subnet = Ipv4Cidr::from(Ipv4Network::from(*v4.subnet()));
+    let private_ip = (*v4.ip()).into();
+    let external_ips = build_external_ipv4_config(
+        source_nat,
+        ephemeral_ip,
+        floating_ips,
+        is_ipv4_only,
+    )?;
+    Ok(Ipv4Cfg { vpc_subnet, private_ip, gateway_ip, external_ips })
+}
+
+fn build_opte_ipv6_config(
+    v6: &PrivateIpv6Config,
+    source_nat: Option<&SourceNatConfig>,
+    ephemeral_ip: Option<&IpAddr>,
+    floating_ips: &[IpAddr],
+    is_ipv6_only: bool,
+) -> Result<Ipv6Cfg, Error> {
+    let gateway_ip = v6
+        .subnet()
+        .iter()
+        // The 0th address is the network address, and OPTE's
+        // gateway sits at the next. See
```

Comment:
Can this call `v6.opte_gateway()` instead?

Reply:
Doh! Yeah, I added those methods and never used them. Thanks!